fyrox-impl/src/material/shader/standard/terrain.shader

```diff
@@ -206,8 +217,8 @@
 
                     outColor = diffuseColor * texture(diffuseTexture, tc);
 
-                    vec4 n = normalize(texture(normalTexture, tc) * 2.0 - 1.0);
-                    outNormal = vec4(normalize(tangentSpace * n.xyz) * 0.5 + 0.5, 1.0);
```

Comment:
Do not remove `* 0.5 + 0.5` - it is needed because normals in the GBuffer are stored in RGBA8 texture, which has 0..255 value range. When you do `* 0.5 + 0.5` you shifting `-1.0..1.0` range to `0.0..1.0` and then it will be unpacked back when reading from the GBuffer.

Reply:
I have done some experiments to see the effect of removing the * 0.5 + 0.5. At first I found nothing apparent, but by greatly increasing the intensity of the light and flattening the terrain, the problem became dramatically visible. It is disturbing that I didn't not see the problem immediately. Now I am dubious of every shadow, questioning whether it is as it should be or whether it is a glitch. I hope people better at shaders than I continue to work on this.